base/arraymath.jl

```diff
@@ -35,53 +35,71 @@ function !(A::AbstractArray{Bool})
 end
 
 ## Binary arithmetic operators ##
-@pure promote_array_type{S<:Number, A<:AbstractArray}(F, ::Type{S}, ::Type{A}) =
-    promote_array_type(F, S, eltype(A), promote_op(F, S, eltype(A)))
-@pure promote_array_type{S<:Number, A<:AbstractArray}(F, ::Type{A}, ::Type{S}) =
-    promote_array_type(F, S, eltype(A), promote_op(F, eltype(A), S))
-
-@pure promote_array_type{S, A, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Real, A<:AbstractFloat, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = A
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F::typeof(./), ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{A}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, A<:Integer, P}(F, ::Type{S}, ::Type{A}, ::Type{P}) = A
-@pure promote_array_type{S<:Integer, P}(F::typeof(./), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F::typeof(.\), ::Type{S}, ::Type{Bool}, ::Type{P}) = P
-@pure promote_array_type{S<:Integer, P}(F, ::Type{S}, ::Type{Bool}, ::Type{P}) = P
+
+@pure _promote_type(cond, T) = cond ? T : (T <: Complex ? Complex{Int} : Int)
+function promote_array_type{S<:Number, A<:AbstractArray}(F, ::Type{S}, ::Type{A})
+    R = eltype(A)
+    cond = isleaftype(R)
+    # If the Array is of a non-concrete type, try to preserve it
+    P = promote_op(F, S, _promote_type(cond, R))
+    T = cond ? P : typejoin(R, S, P)
+    return promote_array_type(F, S, R, T)
+end
+function promote_array_type{S<:Number, A<:AbstractArray}(F, ::Type{A}, ::Type{S})
+    R = eltype(A)
+    cond = isleaftype(R)
+    # If the Array is of a non-concrete type, try to preserve it
+    P = promote_op(F, _promote_type(cond, R), S)
+    T = cond ? P : typejoin(R, S, P)
+    return promote_array_type(F, S, R, T)
+end
+
+@pure promote_array_type(F, ::Type, ::Type, T::Type) = T
+@pure promote_array_type{S<:Real, A<:AbstractFloat}(F, ::Type{S}, ::Type{A}, ::Type) = A
+@pure promote_array_type{S<:Integer, A<:Integer}(F, ::Type{S}, ::Type{A}, ::Type) = A
+@pure promote_array_type{S<:Integer, A<:Integer}(::typeof(./), ::Type{S}, ::Type{A}, T::Type) = T
+@pure promote_array_type{S<:Integer, A<:Integer}(::typeof(.\), ::Type{S}, ::Type{A}, T::Type) = T
+@pure promote_array_type{S<:Integer}(::typeof(./), ::Type{S}, ::Type{Bool}, T::Type) = T
+@pure promote_array_type{S<:Integer}(::typeof(.\), ::Type{S}, ::Type{Bool}, T::Type) = T
+@pure promote_array_type{S<:Integer}(F, ::Type{S}, ::Type{Bool}, T::Type) = T
```

Comment:
there's no value in marking these `pure`. it's obvious that they return a single constant